hydra-node/src/Hydra/HeadLogic.hs

```diff
@@ -382,6 +401,214 @@ deriving instance (IsTx tx, IsChainState tx) => FromJSON (Outcome tx)
+          }
+  ( Open ost@OpenState{coordinatedHeadState = coordinatedHeadState@CoordinatedHeadState{allTxs}}
+    , TrackTxInState{tx}
+    ) ->
+      Open
+        ost
+          { coordinatedHeadState = coordinatedHeadState{allTxs = Map.insert (txId tx) tx allTxs}
+          }
+  ( Open ost@OpenState{coordinatedHeadState = coordinatedHeadState@CoordinatedHeadState{seenTxs}}
+    , NewTxReceived{tx, utxo}
+    ) ->
+      Open
+        ost
+          { coordinatedHeadState =
+              coordinatedHeadState
+                { seenTxs = seenTxs <> [tx]
+                , seenUTxO = utxo
+                }
+          }
+  (Open ost@OpenState{coordinatedHeadState}, ReqSnReceived{nextSnapshot, seenTxs, seenUTxO}) ->
+    Open
+      ost
+        { coordinatedHeadState =
+            coordinatedHeadState
+              { seenSnapshot = SeenSnapshot nextSnapshot mempty
+              , seenTxs
+              , seenUTxO
+              }
+        }
+  (Open ost@OpenState{coordinatedHeadState = chs@CoordinatedHeadState{allTxs}},
+   AckSnConfirmed{snapshot, multisig, confirmed}) ->
+    Open
+      ost
+        { coordinatedHeadState =
+            chs
+              { allTxs = foldr Map.delete allTxs confirmed
+              , confirmedSnapshot =
+                  ConfirmedSnapshot
+                    { snapshot
+                    , signatures = multisig
+                    }
+              , seenSnapshot = LastSeenSnapshot (number snapshot)
+              }
+        }
+  (Open ost@OpenState{coordinatedHeadState}, AckSnPending{snapshot, sigs}) ->
+    Open
+      ost
+        { coordinatedHeadState =
+            coordinatedHeadState
+              { seenSnapshot = SeenSnapshot snapshot sigs
+              }
+        }
+  ( Open OpenState{headId, parameters, coordinatedHeadState = CoordinatedHeadState{confirmedSnapshot}}
+    , HeadClosed{contestationDeadline, newChainState}
+    ) ->
+      Closed
+        ClosedState
+          { parameters
+          , confirmedSnapshot
+          , contestationDeadline
+          , readyToFanoutSent = False
+          , chainState = newChainState
+          , headId
+          }
+  (Closed _, HeadFannedOut{newChainState}) ->
+    Idle IdleState{chainState = newChainState}
+  (Closed cst, ReadyToFanoutReceived{}) ->
+    Closed cst{readyToFanoutSent = True}
+  (Open ost, TickReceived{currentSlot}) ->
+    Open ost{currentSlot}
+  ( Open ost@OpenState{coordinatedHeadState}
+    , SnapshotEmited{requested, lastSeenSnapshot}
+    ) ->
+      Open
+        ost
+          { coordinatedHeadState =
+              coordinatedHeadState
+                { seenSnapshot =
+                    RequestedSnapshot
+                      { lastSeen = seenSnapshotNumber lastSeenSnapshot
+                      , requested
+                      }
+                }
+          }
+  (_, RolledBack{rolledBackChainState}) ->
+    setChainState rolledBackChainState st
+  _ -> st
```

Comment:
This function should be complete: an event should always be applicable. The decision to do something or not has already been taken in the other `update` function and now it's just about applying an event which already happened and can not be wrong.